ColorMap and ValueMap can have multiple sinks

azeno opened the pull request:
# Summary
The PR lifts the restriction that nodes feeding values or resources into the shader graph could only be connected to one sink.

# Details
My initial approach where the sources kept a weak reference on the parameter collections of the sink didn't work because on disconnect the sink was still alive and therefor the source would still write into it.

The approach now is to provide a subscription object (a `CompositeDisposable`) while generating the shader graph (through the `ShaderGeneratorContext.Tags` property) where leaves can tie their value subscription to the lifetime of the entire shader graph.

# Example
The two links from the ColorMap node work as expected:
![image](https://user-images.githubusercontent.com/573618/115959682-b41eb000-a50d-11eb-9401-be92712dd3b0.png)

# Related issues
#323

tebjan commented:
this looks quite good, my only concern is whether we should merge such a change into the stable branch. I think it makes more sense to add it to the preview-2021.4 branch, which most users actively test.

and this change should also be integrated into the ShaderFX node factory, which doesn't exist in the 2021.3 branch.

azeno commented:
I rebased on 2021.4 branch and added the changes to the ShaderFX nodes as well. However my only test was with the TextureFX fader overview patch connecting all effects to the same fader InputValue node. That works until you disconnect in which case the effect doesn't seem to get updated and continues to execute the previous connected InputValue instead going to its default value. But maybe that issue is completely unrelated?

And what would be a good patch to test the key generation thing you mentioned in the chat?

azeno commented:
Another update which I believe is cleaner as it doesn't rely on weak references anymore and the lifetime of the value subscriptions per sink is clearly defined. The overview patch works nicely already. However didn't find the time to cleanup so consider this a WIP.

tebjan commented:
why do you think that a weak reference is so bad? the new approach seems to force every source of an effect or material to implement this mechanism. so even the stride serializer would need to know about this, or not?

this reminds me also of the parent managers in the entities that now force every entity that wants to enter the scene graph to have one of them. I think that was a design decision that forces work on everyone and it requires knowledge that is hidden. in this case, it is less visible to the user, but i would prefer something that stays confined to the area of shader fx and no other system needs to know about it.

azeno commented:
Problem was that the "connection" was still there even after deleting a link. So the source was still writing to the sink.

Stride serializer: In that scenario I'd load a material from disk right? Well then there're no nodes of ours involved which would feed values into the material.

azeno commented:
No longer using subjects. Added link to this PR in case one needs to understand the topic in another lifetime. Would like to merge it back.

tebjan commented on packages/VL.Stride.Runtime/src/Shaders/ShaderGraph.cs:
```diff
@@ -16,11 +16,18 @@
+        /// Gives access to the <see cref="CompositeDisposable"/> whose lifetime is tied to the graph being built.
+        /// See https://github.com/vvvv/VL.Stride/pull/451 for details.
+        /// </summary>
+        internal static readonly PropertyKey<CompositeDisposable> GraphSubscriptions = new PropertyKey<CompositeDisposable>(nameof(GraphSubscriptions), typeof(ShaderGraph));
```
maybe the name `GraphSubscritions `could be more specific? and this tag can be used for anything, I guess, not just our sink subscriptions... it could be called something like `EffectLifetime` or `EffectDisposable`?

azeno replied:
Not sure - are those names really better? Will it always be an effect / effect instance constructing a graph? Or might it be something else? Well I guess we can come up with a better name any time - the property is internal, so one can do a rename anytime without breaking anything.

tebjan replied:
> Not sure - are those names really better?

definitely not, was hope you have something better in mind... :)

tebjan replied:
minor, marking as resolved.

tebjan reviewed (commented):
great work, just one thing from my side: as I wrote in the comment of the parameter setter, it is now optimized for a somewhat rare special case. the default case in the update loop per pin should/could have its own code path, IMO.

azeno commented on packages/VL.Stride.Runtime/src/Shaders/ParameterUpdater.cs:
```diff
@@ -0,0 +1,107 @@
+                if (!comparer.Equals(value, Value))
+                {
+                    this.value = value;
+                    foreach (var (parameters, key) in trackedCollections.Keys)
```
It's not getting constructed in each call - it's getting constructed once, see https://github.com/microsoft/referencesource/blob/master/mscorlib/system/collections/generic/dictionary.cs#L138

Also the foreach is not allocating in this case, see https://github.com/microsoft/referencesource/blob/master/mscorlib/system/collections/generic/dictionary.cs#L812 returning a struct (https://github.com/microsoft/referencesource/blob/master/mscorlib/system/collections/generic/dictionary.cs#L922).

So iterating over one or the other doesn't make a difference, the enumerators in its core do a `dictionary.entries[index]`.

Optimizing for the one parameter collection case introduces will introduce more state but let's see..

azeno replied:
First time in my life pressed that "resolve conversation" button. Point being, I've updated the PR with another commit where the code is optimized for the single parameter collection case.

azeno commented:
Just pushed another commit optimizing for the single parameter collection case. Will keep it as a separate commit to make it easier to see the changes.

tebjan commented on packages/VL.Stride.Runtime/src/Shaders/ParameterUpdater.cs:
```diff
@@ -34,20 +40,37 @@ public TValue Value
                         Upload(parameters, key, ref value);
+                    }
+
+                    if (trackedCollections != null)
```
shouldn't this be an `else if`? it is either the single or many case?

azeno replied:
Assuming the cases are exclusive yes, but what you're asking for below would mean the cases are not we could end up with a mix.

tebjan replied:
yes, I think they should be mutually exclusive and the decision is made on subscribe/unsubscribe. there is could be one bool flag that decides the code path, which would most likely be optimized by the cpu with branch prediction.

tebjan commented on packages/VL.Stride.Runtime/src/Shaders/ParameterUpdater.cs:
```diff
@@ -17,13 +17,19 @@ abstract class ParameterUpdater<TValue, TKey>
         private static readonly EqualityComparer<TValue> comparer = EqualityComparer<TValue>.Default;
 
-        private readonly Dictionary<(ParameterCollection, TKey), RefCountDisposable> trackedCollections = new Dictionary<(ParameterCollection, TKey), RefCountDisposable>();
+        // In most of the cases the parameter collection is known from the start and no other will come into play (pin and effect are in the same node)
```
the shader fx nodes will not know the parameters on creation, they have to wait until the first GenerateShaderSource call. but they are also single-sink most of the time. will they just use another updater or how does the logic work in that case?

and what happens when a source gets removed? will it go back to the single case?

azeno replied:
Currently it's the single case where the collection is known from the start and the multi case where it's about zero to many collections which pop in and out over time.

I fear optimizing the dynamic case might get super tricky and hard to get right. Just to avoid that GetEnumerator call? Shouldn't we strive for correctness first? Then we have a baseline which works. Optimize later as they say.

tebjan replied:
> Shouldn't we strive for correctness first? Then we have a baseline which works. Optimize later as they say.

yes, definitely the right time now while the system is known in the head. if you check the backlog, this will never ever be done in the next 5-10 years.

azeno replied:
Modified so it doesn't matter whether the collection & key combination is known up-front or not.

tebjan replied:
minor, marking as resolved.

tebjan commented on packages/VL.Stride.Runtime/src/Shaders/ParameterUpdater.cs:
```diff
@@ -41,14 +42,14 @@ public TValue Value
                     this.value = value;
 
+                    // Fast code path in case we track only a single combination
                     if (parameters != null)
```
minor nitpick, I would even use a bool flag that is set in subscribe. so the runtime knows a value that almost never changes. but otherwise. looks great!

tebjan replied:
minor, marking as reolved.

tebjan commented:
major PR and ready to merge from my side. definitely, the better UX compared to a warning. should there be an issue after the merge, I would consider it a normal bug.

also, thanks for the detour into the multiple sink problem, I am sure we will have this problem again in the future and we will be able to learn from this solution. as in #282
